rust/src/applayerpgsql/pgsql.rs

```diff
@@ -0,0 +1,699 @@
+                found = true;
+                index = i;
+                break;
+            }
+        }
+        if found {
+            self.transactions.remove(index);
+        }
+    }
+
+    pub fn get_tx(&mut self, tx_id: u64) -> Option<&PgsqlTransaction> {
+        for tx in &mut self.transactions {
+            if tx.tx_id == tx_id + 1 {
+                return Some(tx);
+            }
+        }
+        return None;
+    }
+
+    fn new_tx(&mut self) -> PgsqlTransaction {
+        let mut tx = PgsqlTransaction::new();
+        self.tx_id += 1;
+        tx.tx_id = self.tx_id;
+        return tx;
+    }
+
+    fn find_request(&mut self) -> Option<&mut PgsqlTransaction> {
+        for tx in &mut self.transactions {
+            if tx.response.is_none() {
+                return Some(tx);
+            }
+        }
+        None
+    }
+
+    fn parse_request(&mut self, input: &[u8]) -> AppLayerResult {
+        // We're not interested in empty requests.
+        if input.len() == 0 {
+            return AppLayerResult::ok();
+        }
+
+        // If there was gap, check we can sync up again.
+        if self.request_gap {
+            // TODO I have changed probe to just return a boolean. will this be an issue, later on?
+            if !probe_ts(input) {
+                // The parser now needs to decide what to do as we are not in sync.
+                // For this pgsql, we'll just try again next time.
+                SCLogNotice!("Suricata thinks there's a gap in the request");
+                return AppLayerResult::ok();
+            }
+
+            // It looks like we're in sync with a message header, clear gap
+            // state and keep parsing.
+            self.request_gap = false;
+        }
+
+        let mut start = input;
+        while start.len() > 0 {
+            match parser::pgsql_parse_request(start) {
+                Ok((rem, request)) => {
+                    start = rem;
+
+                    SCLogNotice!("Request: {:?}", request);
+                    let mut tx = self.new_tx();
+                    tx.request = Some(request);
+                    self.transactions.push(tx);
+                },
+                Err(nom::Err::Incomplete(needed)) => {
+                    // Not enough data. This parser doesn't give us a good indication
+                    // of how much data is missing so just ask for one more byte so the
+                    // parse is called as soon as more data is received.
+                    SCLogNotice!("Suricata thinks request is incomplete");
+                    let consumed = input.len() - start.len();
+                    let needed_estimation = start.len() + 1;
+                    SCLogNotice!("Needed: {:?}, estimated needed: {:?}", needed, needed_estimation);
+                    return AppLayerResult::incomplete(consumed as u32, needed_estimation as u32);
+                },
+                Err(_) => {
+                    return AppLayerResult::err();
```

Comment:
Should we set an event in this case ? and have decoder-event rules in the `rules` directory ?

cf `git grep -i mqtt | cut -d: -f1 | uniq` to see which files need a change for a new app-layer protocol

Reply:
like `rules/mqtt-events.rules` ?